File: frameworks/python/polars/workloads.py
import polars as pl

def project(df: pl.DataFrame):
    return df.lazy().select(["state", "votes"]).collect()


def group_count(df: pl.DataFrame):
    return df.lazy().groupby("party").agg(pl.count("unique_col")).collect()


def sort(df: pl.DataFrame):
    return df.lazy().sort(by="votes", reverse=True).collect()


def flip_bool(df: pl.DataFrame):
    return df.lazy().select([pl.col("won").is_not()]).collect()


def find_max(df: pl.DataFrame):
    return df.lazy().select("votes").max().collect()


def select(df: pl.DataFrame):
    return df.lazy().filter((pl.col("party") == "DEM") & (pl.col("votes") > 20000) & pl.col("won")).collect()


def select_group_mean(df: pl.DataFrame):
    return df.lazy().filter(pl.col("won")).groupby("party").agg(pl.col("votes").mean()).collect()


def longest_string_in_column(df: pl.DataFrame):
    return df.lazy().select([pl.col("candidate").apply(len).max()]).collect()


def join(df: pl.DataFrame):
    return df.lazy().join(df.lazy(), on="unique_col", how="inner").collect()


def select_p90_cols(df: pl.DataFrame):
    return df[:, 1: round(df.shape[1] * 0.9)]


def select_p90_rows(df: pl.DataFrame):
    return df[1: round(df.shape[0] * 0.9)]


def combine_cols(df: pl.DataFrame):
    return df\
        .lazy()\
        .with_columns([pl
                      .concat_str(["candidate", "party", "won"], sep=" ")
                      .alias("combined")]).collect()

File: frameworks/r/dplyr/workloads.R
library(dplyr)

project <- function(df) {
  return(dplyr::select(df, state, votes))
}

group_count <- function(df){ 
  return(df %>% dplyr::group_by(votes) %>% count())
}

sort <- function(df){
  return(arrange(df, desc(votes)))
}

flip_bool <- function(df){
  return(summarise(df, won = !won))
}

find_max <- function(df){
  return(summarise(df, max(votes)))
}

select <- function(df){
  return(filter(df, party == "DEM", won == TRUE, votes > 20000))
}

select_group_mean <- function(df){
  return(
    df %>% filter(won == TRUE) %>% group_by(party) %>% summarise(votes = mean(votes))
  )
}

longest_string_in_column <- function(df) {
  return(df %>% summarise(candidate = nchar(candidate)) %>% arrange(desc(candidate)))
}

join <- function(df){
  return(df %>% inner_join(df, by="unique_col"))
}

select_p90_cols <- function(df){
  return(dplyr::select(df, c(1, round(ncol(df) * 0.9))))
}

select_p90_rows <- function(df){
  return(slice(df, c(1, round(nrow(df) * 0.9))))
}

combine_cols <- function(df){
  df$combined = paste(df$candidate, df$party, df$won)
  return(df)
}

